openprescribing/pipeline/management/commands/outlier_reports.py

```diff
@@ -0,0 +1,1656 @@
+            "from_date": from_date.strftime(MakeHtml.REPORT_DATE_FORMAT),
+            "to_date": to_date.strftime(MakeHtml.REPORT_DATE_FORMAT),
+            "entity_type": entity_type,
+            "entity_code": entity_code,
+        }
+
+        with open(output_path, "w") as f:
+            f.write(template.render(context))
+
+
+class DatasetBuild:
+    """
+    Calls outlier dataset building stored procedure on bigquery
+    fetches, and encapsulates results of this process.
+
+    Attributes
+    ----------
+    from_date : datetime.date
+        start date of outlier reporting period
+    to_date : datetime.date
+        end date of outlier reporting period
+    n_outliers : int
+        number of outliers to include in each "high" and "low" outtlier set
+    entities : List[str]
+        list of column names for entity types to report e.g. "ccg"
+    force_rebuild : bool
+        force rebuilding of outlier dataset within bigquery and rebuilding
+        of local data caches.
+    numerator_column : str
+        column name for numerator values in ratio calculation
+        N.B: not yet integrated into bigquery stored procedure
+    denominator_column : str
+        column name for denominator values in ratio calculation
+        N.B: not yet integrated into bigquery stored procedure
+    """
+
+    # consts
+    _DATEFMT = "%Y-%m-%d"
+    _KNOWN_ENTITIES = ["practice", "ccg", "pcn", "stp"]
+
+    def __init__(
+        self,
+        from_date: date,
+        to_date: date,
+        n_outliers: int,
+        entities: List[str],
+        force_rebuild: bool = False,
+        numerator_column: str = "chemical",
+        denominator_column: str = "subpara",
+    ) -> None:
+        assert (
+            type(to_date) == date and type(from_date) == date
+        ), "date args must be dates"
+        assert to_date > from_date, "to date must be after from date"
+        self.from_date = from_date
+        self.to_date = to_date
+
+        assert n_outliers > 0, "n must be greater than zero"
+        self.n_outliers = n_outliers
+
+        assert len(entities) > 0, "list of entities must be populated"
+        for e in entities:
+            assert e in self._KNOWN_ENTITIES, f"{e} not recognised entity"
+        self.entities = entities
+
+        self.force_rebuild = force_rebuild
+        self.numerator_column = numerator_column
+        self.denominator_column = denominator_column
+
+        self.results: Dict[str, pd.DataFrame] = {}
+        self.results_items: Dict[str, pd.DataFrame] = {}
+        self.results_measure_arrays: Dict[str, pd.DataFrame] = {}
+        self.entity_hierarchy: Dict[str, Dict[str, List[str]]] = {}
+
+    def run(self) -> None:
+        """
+        Execute outlier dataset build stored procedure
+
+        Populates build_id attribute upon completion
+        """
+        sql = f"""
+            CALL `ebmdatalab.outlier_detection.build_outliers`(

```

Comment:
Is the code for this stored procedure checked in anywhere?

Reply:
https://github.com/ebmdatalab/outliers/tree/master/data/static_outlier_sql